### PR TITLE
Add CurrentAssetDeck

### PR DESCRIPTION
**Pull request by UsamaEquinorAFK**

### CODE REVIEW

File: backend/api/Controllers/RobotController.cs [hunk 1]
```diff
@@ -643,34 +643,40 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+    [Route("start-localization")]
     [ProducesResponseType(typeof(Mission), StatusCodes.Status200OK)]
     [ProducesResponseType(StatusCodes.Status400BadRequest)]
     [ProducesResponseType(StatusCodes.Status401Unauthorized)]
     [ProducesResponseType(StatusCodes.Status403Forbidden)]
     [ProducesResponseType(StatusCodes.Status404NotFound)]
     [ProducesResponseType(StatusCodes.Status500InternalServerError)]
     public async Task<ActionResult<Mission>> StartLocalizationMission(
-        [FromRoute] string robotId,
-        [FromBody] Pose localizationPose
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:655]: Don't remove localizationPose as input. It's used in https://github.com/equinor/flotilla/pull/870

**Comment by Eddasol** [comment]
Rewrite commit message to "Add CurrentAssetDeck"

File: backend/api/Controllers/RobotController.cs (hunk 1 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs:655]: I added this in the new class ScheduleLocalizationMissionQuery.cs

File: backend/api/Controllers/RobotController.cs [hunk 2]
```diff
@@ -686,7 +693,7 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
         IsarMission isarMission;
         try
         {
-            isarMission = await _isarService.StartLocalizationMission(robot, localizationPose);
+            isarMission = await _isarService.StartLocalizationMission(robot, deck.DefaultLocalizationPose);
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:696]:
```suggestion
            isarMission = await _isarService.StartLocalizationMission(robot, scheduleLocalizationMissionQuery.localizationPose);
```

File: backend/api/Controllers/RobotController.cs [hunk 3]
```diff
@@ -643,34 +643,41 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
             );
             return Conflict($"The Robot is not available ({robot.Status})");
         }
+        var deck = await _assetDeckService.ReadById(scheduleLocalizationMissionQuery.DeckId);
+
+        if (deck == null)
+        {
+            _logger.LogWarning("Could not find deck with id={id}", scheduleLocalizationMissionQuery.DeckId);
+            return NotFound("Deck not found");
+        }
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:680]: This can be removed

File: backend/api/EventHandlers/MqttEventHandler.cs [hunk 4]
```diff
@@ -236,15 +236,21 @@ private static void UpdateHostIfChanged(string host, ref Robot robot, ref List<s
...
             updatedFields.Add($"\nCurrentAsset ({robot.CurrentAsset} -> {newCurrentAsset})\n");
             robot.CurrentAsset = newCurrentAsset;
-        }
 
+            if (robot.CurrentAssetDeck != null)
+            {
+                return;
+            }
+
+        }
```
Comment (Eddasol) [inline backend/api/EventHandlers/MqttEventHandler.cs:253]: this part doesn't do anything

File: backend/api/EventHandlers/MqttEventHandler.cs [hunk 5]
```diff
@@ -236,15 +236,21 @@ private static void UpdateHostIfChanged(string host, ref Robot robot, ref List<s
         private static void UpdateCurrentAssetIfChanged(string newCurrentAsset, ref Robot robot,
             ref List<string> updatedFields)
         {
+
```
Comment (Eddasol) [inline backend/api/EventHandlers/MqttEventHandler.cs:239]: Remove new line

File: backend/api/Database/Models/RobotModel.cs [hunk 6]
```diff
@@ -107,7 +107,9 @@ public void UpdateAverageDurationPerTag(List<Mission> recentMissionsForModelType
                             )
                 )
                 .ToList();
-
+            // If no valid task times, return
+            if (timeSpentPerTask.All(time => time < 0))
+                return;
```
Comment (Eddasol) [inline backend/api/Database/Models/RobotModel.cs:112]: Add new line before and after this for better readability

File: backend/api/Controllers/RobotController.cs [hunk 7]
```diff
@@ -707,13 +714,16 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
             return StatusCode(StatusCodes.Status500InternalServerError, message);
         }
 
+
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:717]: Remove double new line

File: backend/api/Controllers/RobotController.cs [hunk 8]
```diff
@@ -707,13 +714,16 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
         mission.UpdateWithIsarInfo(isarMission);
         mission.Status = MissionStatus.Ongoing;
 
         await _missionService.Create(mission);
 
+
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:723]: Remove double new line

File: backend/api/EventHandlers/MqttEventHandler.cs (hunk 4 above)
Comment (UsamaEquinorAFK) [reply backend/api/EventHandlers/MqttEventHandler.cs:253]: Nope its not - just kept it incase we add a few lines of code related to currentassetdeck below it which we might need to skip if we have CurrentAssetDeck as null

File: backend/api/Controllers/RobotController.cs (hunk 3 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs:680]: It is used below at robot.CurrentAssetDeck = deck

Comment (Eddasol) [reply backend/api/Controllers/RobotController.cs:680]: Right, didn't see that one :)

File: backend/api/EventHandlers/MqttEventHandler.cs (hunk 4 above)
Comment (Eddasol) [reply backend/api/EventHandlers/MqttEventHandler.cs:252]: It better to add it when we need it instead of adding code that we might need some day

**Comment by Eddasol** [comment]
> bell Changes in database folder detected bell Do these changes require **adding new migrations**? thinking In that case follow [these steps](https://github.com/equinor/flotilla/tree/main/backend#Database-model-and-EF-Core). If you are uncertain, ask a database admin on the team smile

Remember to add migrations

File: backend/api/Controllers/RobotController.cs (hunk 8 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs:715]: Fixed now :)

File: backend/api/Controllers/RobotController.cs (hunk 2 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs]: This is added now :)

File: backend/api/EventHandlers/MqttEventHandler.cs (hunk 4 above)
Comment (UsamaEquinorAFK) [reply backend/api/EventHandlers/MqttEventHandler.cs]: Done :)

File: backend/api/EventHandlers/MqttEventHandler.cs [hunk 9]
```diff
@@ -236,16 +236,19 @@ private static void UpdateHostIfChanged(string host, ref Robot robot, ref List<s
         private static void UpdateCurrentAssetIfChanged(string newCurrentAsset, ref Robot robot,
             ref List<string> updatedFields)
         {
-            if (newCurrentAsset.Equals(robot.CurrentAsset, StringComparison.Ordinal))
+            if (robot.CurrentAssetDeck != null)
```
[the remark this answers is not in the document]
Comment (UsamaEquinorAFK) [reply backend/api/EventHandlers/MqttEventHandler.cs]: Agreed - removed this check for currentAssetdeck

File: backend/api/Controllers/RobotController.cs [hunk 10]
```diff
@@ -709,16 +709,27 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+        var deck = await _assetDeckService.ReadById(scheduleLocalizationMissionQuery.DeckId);
+
+        if (deck == null)
+        {
+            _logger.LogWarning("Could not find deck with id={id}", scheduleLocalizationMissionQuery.DeckId);
+            return NotFound("Deck not found");
+        }
+
+        deck.DeckName = scheduleLocalizationMissionQuery.DeckId;
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:725]: Why do we want the deck name to be the deck id?

File: backend/api/Controllers/RobotController.cs [hunk 11]
```diff
@@ -709,16 +709,27 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+        deck.AssetCode = robot.CurrentAsset;
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:726]: Opposite. We don't want to  make changes to the assetDeck

File: backend/api/Controllers/RobotController.cs [hunk 12]
```diff
@@ -709,16 +709,27 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+        deck.DefaultLocalizationPose = scheduleLocalizationMissionQuery.LocalizationPose;
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:727]: Do not change this. Just because we want the localizationPose to be in the middle of the module once, doesn't mean that we always want it to be there

File: backend/api/EventHandlers/MqttEventHandler.cs [hunk 13]
```diff
@@ -243,8 +243,8 @@ private static void UpdateHostIfChanged(string host, ref Robot robot, ref List<s
 
             updatedFields.Add($"\nCurrentAsset ({robot.CurrentAsset} -> {newCurrentAsset})\n");
             robot.CurrentAsset = newCurrentAsset;
-        }
 
+        }
```
Comment (Eddasol) [inline backend/api/EventHandlers/MqttEventHandler.cs:247]: Revert this change

File: backend/api/Controllers/RobotController.cs (hunk 11 above)
Comment (Eddasol) [reply backend/api/Controllers/RobotController.cs:726]: Would make more sense to set
robot.CurrentAsset = deck.AssetCode

File: backend/api/Controllers/RobotController.cs (hunk 12 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs]: Done :)

File: backend/api/Controllers/RobotController.cs (hunk 11 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs]: Done :)

File: backend/api/Controllers/RobotController.cs (hunk 10 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs]: Done :)

File: backend/api/EventHandlers/MqttEventHandler.cs (hunk 13 above)
Comment (UsamaEquinorAFK) [reply backend/api/EventHandlers/MqttEventHandler.cs:246]: Done :)

**Review by eivindsjovold — DISMISSED** [review]

File: frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx [hunk 14]
```diff
@@ -97,8 +97,7 @@ export const LocalizationDialog = ({ robot }: RobotProps): JSX.Element => {
 
     const onClickLocalize = () => {
         if (selectedAssetDeck && localisationPose) {
-            BackendAPICaller.postLocalizationMission(localisationPose, robot.id)
-        }
```
[the remark this answers is not in the document]
Comment (UsamaEquinorAFK) [reply frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx:101]: Done. Thanks !

[the remark this answers is not in the document]
Comment (UsamaEquinorAFK) [reply frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx:101]: @eivindsjovold  can you please review again since you chose to request changes and not simply review, now you need to review until approved for this to go in. Thanks !

File: backend/api/Controllers/RobotController.cs [hunk 15]
```diff
@@ -709,16 +709,24 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
         robot.Status = RobotStatus.Busy;
         robot.CurrentMissionId = mission.Id;
+
+        var deck = await _assetDeckService.ReadById(scheduleLocalizationMissionQuery.DeckId);
+
+        if (deck == null)
+        {
+            _logger.LogWarning("Could not find deck with id={id}", scheduleLocalizationMissionQuery.DeckId);
+            return NotFound("Deck not found");
+        }
```
Comment (Eddasol) [inline backend/api/Controllers/RobotController.cs:723]: Consider moving this before running
"isarMission = await _isarService.StartLocalizationMission(robot, scheduleLocalizationMissionQuery.LocalizationPose);"
such that we cancel the localization if we cant find the correct deck

Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs]: Done !

**Review by Eddasol — APPROVED** [review]
LGTM

File: backend/api/Services/RobotService.cs [hunk 16]
```diff
@@ -30,7 +30,7 @@ public RobotService(FlotillaDbContext context)
 
         private IQueryable<Robot> GetRobotsWithSubModels()
         {
-            return _context.Robots.Include(r => r.VideoStreams).Include(r => r.Model);
+            return _context.Robots.Include(r => r.VideoStreams).Include(r => r.Model).Include(r => r.CurrentAssetDeck);
```
Comment (eivindsjovold) [inline backend/api/Services/RobotService.cs:33]: should this be added in flotilladbcontext?

Comment (UsamaEquinorAFK) [reply backend/api/Services/RobotService.cs:33]: In my understanding this is the place where we are including sub models in the context object and inside FlotilladB we add constraints on the objects of these sub models. Atleast this is what I think happens with the Class Assetdeck. I would keep these two things seperate for clarity.

Comment (eivindsjovold) [reply backend/api/Services/RobotService.cs:33]: I think context definition should live in the db context - currentassetdeck is always a property that lives on the robot model if I have not misunderstood

File: backend/api/Controllers/RobotController.cs [hunk 17]
```diff
@@ -643,30 +643,30 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+        [FromBody] ScheduleLocalizationMissionQuery scheduleLocalizationMissionQuery
+
     )
     {
-        var robot = await _robotService.ReadById(robotId);
+        var robot = await _robotService.ReadById(scheduleLocalizationMissionQuery.RobotId);
         if (robot == null)
         {
-            _logger.LogWarning("Could not find robot with id={id}", robotId);
+            _logger.LogWarning("Could not find robot with id={id}", scheduleLocalizationMissionQuery.RobotId);
```
Comment (eivindsjovold) [inline backend/api/Controllers/RobotController.cs:661]: Does this actually log the id? is id a variable here?

File: backend/api/Controllers/RobotController.cs [hunk 18]
```diff
@@ -643,30 +643,30 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
             return NotFound("Robot not found");
         }
 
         if (robot.Status is not RobotStatus.Available)
         {
             _logger.LogWarning(
                 "Robot '{id}' is not available ({status})",
```
Comment (eivindsjovold) [inline backend/api/Controllers/RobotController.cs:668]: same as above

File: backend/api/Controllers/RobotController.cs [hunk 19]
```diff
@@ -683,10 +683,18 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
             MapMetadata = new MapMetadata()
         };
 
+        var deck = await _assetDeckService.ReadById(scheduleLocalizationMissionQuery.DeckId);
+
+        if (deck == null)
+        {
+            _logger.LogWarning("Could not find deck with id={id}", scheduleLocalizationMissionQuery.DeckId);
```
Comment (eivindsjovold) [inline backend/api/Controllers/RobotController.cs:690]: same as above

File: backend/api/Database/Models/Robot.cs [hunk 20]
```diff
@@ -28,6 +28,8 @@ public class Robot
 
         public string CurrentAsset { get; set; }
 
+        public AssetDeck? CurrentAssetDeck { get; set; }
```
Comment (eivindsjovold) [inline backend/api/Database/Models/Robot.cs:31]: Is this nullable?

File: frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx [hunk 21]
```diff
@@ -97,11 +97,10 @@ export const LocalizationDialog = ({ robot }: RobotProps): JSX.Element => {
 
     const onClickLocalize = () => {
         if (selectedAssetDeck && localisationPose) {
-            BackendAPICaller.postLocalizationMission(localisationPose, robot.id)
+            BackendAPICaller.postLocalizationMission(localisationPose, robot.id, selectedAssetDeck.id)
```
Comment (eivindsjovold) [inline frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx:100]: Make a corresponding structure in frontend as in backed for ScheduleLocalizationMissionQuery

File: backend/api/Services/RobotService.cs (hunk 16 above)
Comment (UsamaEquinorAFK) [reply backend/api/Services/RobotService.cs:33]: Anyways refactoring of this can be another issue.

File: backend/api/Controllers/RobotController.cs (hunk 17 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs:661]: Yes RobotId is a variable in scheduleLocalizationMissionQuery

File: backend/api/Controllers/RobotController.cs (hunk 18 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs:668]: Its a variable in scheduleLocalizationMissionQuery

File: backend/api/Database/Models/Robot.cs (hunk 20 above)
Comment (UsamaEquinorAFK) [reply backend/api/Database/Models/Robot.cs:31]: Yes for now. It will be reintroduced in a different way in a different PR. It can be made non-nullable not sure if it adds any value ?

File: frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx (hunk 21 above)
Comment (UsamaEquinorAFK) [reply frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx:100]: Not sure if its required for this to work. Maybe it is a better approach will look into it.

File: backend/api/Controllers/RobotController.cs (hunk 19 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/RobotController.cs:690]: Answered in the above ones.

File: frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx (hunk 21 above)
Comment (UsamaEquinorAFK) [reply frontend/src/components/Pages/RobotPage/LocalizationDialog.tsx:100]: Would make more sense if the value was fetched from backend and sent to frontend unlike this case where it is the other way around. Dont you agree. ? :p

File: backend/api/Database/Models/Robot.cs (hunk 20 above)
Comment (UsamaEquinorAFK) [reply backend/api/Database/Models/Robot.cs:31]: It should be not required incase we make an instance of robot when it is not localised yet, because then we have to make a dummy key and key is supposed to be unique. Key is a variable inside AssetDeck class.

**Review by andchiind — APPROVED** [review]
LGTM
